src/nftables.c

```diff
@@ -0,0 +1,371 @@
+
+  struct nftables_rule *rule_info;
+  struct nftables_rule **list;
+  int family;
+  char *fields[3] = {0};
+
+  if (strcasecmp(key, "ip") == 0) {
+    family = NFPROTO_IPV4;
+  } else if (strcasecmp(key, "ip6") == 0) {
+    family = NFPROTO_IPV6;
+  } else if (strcasecmp(key, "arp") == 0) {
+    family = NFPROTO_ARP;
+  } else if (strcasecmp(key, "bridge") == 0) {
+    family = NFPROTO_BRIDGE;
+  } else if (strcasecmp(key, "netdev") == 0) {
+    family = NFPROTO_NETDEV;
+  } else if (strcasecmp(key, "inet") == 0) {
+    family = NFPROTO_INET;
+  } else {
+    ERROR("Unknown family: %s", key);
+    return -1;
+  }
+
+  char *value_copy = strdup(value);
+  if (value_copy == NULL) {
+    ERROR("strdup failed: %s", STRERRNO);
+    return 1;
+  }
+
+  int fields_num = strsplit(value_copy, fields, 3);
+  if (fields_num < 2) {
+    free(value_copy);
+    return 1;
+  }
+
+  char *table = fields[0];
+  char *chain = fields[1];
+  char *comment = "";
+
+  if (fields_num == 3) {
+    comment = fields[2];
+  } else if (fields_num > 3) {
+    plugin_log(LOG_NOTICE, "Ignoring excess arguments");
+  }
+
+  rule_info = (struct nftables_rule *)malloc(sizeof(struct nftables_rule));
+  if (!rule_info) {
+    ERROR("nftables plugin: nftables rule_info malloc failed: %s", STRERRNO);
+    return 1;
+  } else {
+    rule_info->family = family;
+    rule_info->table = strdup(table);
+    rule_info->chain = strdup(chain);
+    rule_info->comment = strdup(comment);
+  }
+  free(value_copy);
+
+  list = realloc(rule_list, (rule_count + 1) * sizeof(struct nftables_rule *));
+  if (list == NULL) {
+    ERROR("nftables plugin: rule_info list realloc failed: %s", STRERRNO);
+    return 1;
+  }
+
+  rule_list = list;
+  rule_list[rule_count] = rule_info;
+  rule_count++;
+
+  plugin_log(LOG_INFO,
+             "Stored %s rule info -> table: %s, chain: %s, comment: %s", key,
+             rule_info->table, rule_info->chain, rule_info->comment);
+
+  return 0;
+} /* int nftables_config */
+
+static int nftables_init(void) {
+  nl = mnl_socket_open(NETLINK_NETFILTER);
+
+  plugin_log(LOG_NOTICE, "Initializing nftables plugin...");
+  for (int i = 0; i < rule_count; i++) {
+    struct nftables_rule *r = rule_list[i];
+    plugin_log(LOG_INFO,
+               "rule_list[%d] => family: %d table: %s chain: %s comment: %s", i,
+               r->family, r->table, r->chain, r->comment);
+  }
+
+  if (mnl_socket_bind(nl, 0, MNL_SOCKET_AUTOPID) < 0) {
+    plugin_log(LOG_ERR, "error calling mnl_socket_bind");
+    return -1;
+  }
+
+  portid = mnl_socket_get_portid(nl);
+  plugin_log(LOG_INFO, "mnl socket bind, portid: %d", portid);
+
+  return 0;
+} /* int nftables_init */
+
+static int nftables_read(void) {
+  return nl_match_rules();
+} /* int nftables_read */
+
+static int nftables_shutdown(void) {
+  mnl_socket_close(nl);
+  struct nftables_rule *r;
+
+  for (int i = 0; i < rule_count; i++) {
+    r = rule_list[i];
+    free(r->table);
+    free(r->chain);
+    free(r->comment);
```

Comment:
As of now, a rule filter specify in the plugin section if it does not contain a comment will initialize `r->comment` to `""`

https://github.com/pvxe/collectd/blob/8ced2a0bdcda5731657bac39f399fe270defc803/src/nftables.c#L284-L303

For this, `table`, `chain` and `comment` are initialized and are to be freed for each entry in the list when shutdown.